pqarrow/arrowutils/sort.go

```diff
@@ -0,0 +1,105 @@
+package arrowutils
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sort"
+
+	"github.com/apache/arrow/go/v14/arrow"
+	"github.com/apache/arrow/go/v14/arrow/array"
+	"github.com/apache/arrow/go/v14/arrow/compute"
+	"github.com/apache/arrow/go/v14/arrow/memory"
+	"golang.org/x/exp/constraints"
+)
+
+// SortRecord sorts the given record's rows by the given column. Currently only supports int64, string and binary columns.
+func SortRecord(mem memory.Allocator, r arrow.Record, cols []int) (*array.Int64, error) {
+	if len(cols) > 1 {
+		return nil, fmt.Errorf("sorting by multiple columns isn't implemented yet")
+	}
+	indicesBuilder := array.NewInt64Builder(mem)
+
+	if r.NumRows() == 0 {
+		return indicesBuilder.NewInt64Array(), nil
+	}
+	if r.NumRows() == 1 {
+		indicesBuilder.Append(0)
+		return indicesBuilder.NewInt64Array(), nil
+	}
+
+	indices := make([]int64, r.NumRows())
+	// populate indices
+	for i := range indices {
+		indices[i] = int64(i)
+	}
+
+	switch c := r.Column(cols[0]).(type) {
+	case *array.Int64:
+		sort.Sort(orderedSorter[int64]{array: c, indices: indices})
+	case *array.String:
+		sort.Sort(orderedSorter[string]{array: c, indices: indices})
+	case *array.Binary:
+		sort.Sort(binarySort{array: c, indices: indices})
+	default:
+		return nil, fmt.Errorf("unsupported column type for sorting %T", c)
+	}
+
+	indicesBuilder.AppendValues(indices, nil)
+	return indicesBuilder.NewInt64Array(), nil
+}
+
+// ReorderRecord reorders the given record's rows by the given indices.
+// This is a wrapper around compute.Take which handles the type castings.
+func ReorderRecord(r arrow.Record, indices arrow.Array) (arrow.Record, error) {
+	res, err := compute.Take(
```

Comment:
Awesome, I also wanted to note, this will work majority of the time. We support dictionaries and there is no dictionary kernel for `Take` yet.

We need to optimise

-  If record has no dictionary column. Use `compute.Take` on  hte record
-  If there is a dictionary column. Use `compute.Take` on individual non dictionary columns and fall back to manual taking for the dictionary and then assembling the record (preferably concurrently).


This can be done on a separate PR. Unless I'm mistaken, the PR just adds these functions but there is no expected call site for them in yet( There will be a need for a lot of changes for this to be used in  my logictest PR). 

I'm on mobile now , so I cant write what is needed for the logictest to use these functions.

Reply:
Oh, that's nasty and a really good comment! 
We should make sure this is supported in follow-up PRs!